src/language/i18n.js

```diff
@@ -25,7 +31,7 @@ const getInitialLocale = () => {
 }
 
 async function dynamicActivate(locale) {
-  const { messages } = await import(`@lingui/loader!./locales/${locale}/catalog.json`)
+  const { messages } = await localeFiles[locale]()
   i18n.load(locale, messages)
   i18n.activate(locale)
 }
```

Comment:
why does _setLanguage needs the two cases of `init` ?
why the useEffect cant simply call _setLanguage(language)?

Reply:
The _setLanguage is there for when the user decides to change the current languange. It's not intended to be used inside useEffect.
It uses `init` because since language loading is async, the user may request a language change in both states of loading.

Reply:
i still dont understand. if I follow the logic then:
lets assume useEffect is initially called with 'en'
while it is loading the user choose 'es'
because init is still true it will not call "dynamicActivate" therefor it will not load the spanish language files, therefor user will not see spanish but still english.

Did you encounter any issues when calling i18ln.load multiple times? is this code actually solving some kind of an issue or is it just your assumption?

Reply:
I have not personally encountered any problems with loading multiple times, but the [project I used as a reference](https://github.com/GoodDollar/GoodProtocolUI/blob/master/src/language/index.tsx) is using that code for several months. Should I change the code to simplify _setLanguage?

Reply:
@johnsmith-gooddollar what do you think? I dont see the reason for the current logic

Reply:
i'll simplify it

Reply:
@alecsantos96 your locale detection is **browser only**
it won't work on Android

first of all you had to use AsyncStorage form RN instead of localStorage
second point - you had to use Android or RN tools to get system locale

this need to be updated ASAP!

Reply:
@alecsantos96 please:

1. get latest master (with my fixes) from there https://github.com/GoodDollar/GoodDAPP/commit/8ce0abcc967b4286525114454c004c9b18619610
2. add cross-platform system locale detection ASAP!

Reply:
added some fixes, simpified logic
now it looks pretty simple:

```javascript
const [language, setCurrentLanguage] = useState(null)

  const setLanguage = useCallback(
    async language => {
      await I18n.dynamicActivate(language)
      setCurrentLanguage(language)
    },
    [setCurrentLanguage],
  )

  useEffect(() => {
    I18n.getInitialLocale().then(setLanguage)
  }, [setLanguage])
```